daemon/daemon_main.go

```diff
@@ -1328,6 +1331,26 @@ func runDaemon() {
 		}()
 	}
 
+	// Migrating the ENI datapath must happen before the API is served to
+	// prevent endpoints from being created. It also must be before the health
+	// initialization logic which creates the health endpoint, for the same
+	// reasons as the API being served. We want to ensure that this migration
+	// logic runs before any endpoint creates.
+	if option.Config.IPAM == option.IPAMENI {
+		err, revErr := linuxrouting.MigrateENIDatapath(&interfaceNumberGetter{})
+		switch {
+		case err != nil && revErr != nil:
+			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. " +
+				"Connectivity is not guaranteed and manual intervention is required.")
+		case err != nil && revErr == nil:
+			log.WithError(err).Warning("Failed to migrate ENI datapath. The original datapath has been reinstated. " +
+				"Connectivity has not been affected. Consider investigating the error as it should be temporary and " +
+				"proceed with the migration again.")
+		case err == nil && revErr == nil:
+			log.Info("Migration of ENI datapath successful")
```

Comment:
nit: If I follow right, this will print regardless of whether migration was actually performed.

Reply:
Thanks. Here's the diff to fix this:

```diff --git a/daemon/daemon_main.go b/daemon/daemon_main.go
index 21aeb9133..5c72345b7 100644
--- a/daemon/daemon_main.go
+++ b/daemon/daemon_main.go
@@ -1337,7 +1337,7 @@ func runDaemon() {
 	// reasons as the API being served. We want to ensure that this migration
 	// logic runs before any endpoint creates.
 	if option.Config.IPAM == option.IPAMENI {
-		err, revErr := linuxrouting.MigrateENIDatapath(&interfaceNumberGetter{})
+		migrated, err, revErr := linuxrouting.MigrateENIDatapath(&interfaceNumberGetter{})
 		switch {
 		case err != nil && revErr != nil:
 			log.WithError(err).Fatal("Failed to migrate ENI datapath. The original datapath failed to be reinstated. " +
@@ -1346,8 +1346,8 @@ func runDaemon() {
 			log.WithError(err).Warning("Failed to migrate ENI datapath. The original datapath has been reinstated. " +
 				"Connectivity has not been affected. Consider investigating the error as it should be temporary and " +
 				"proceed with the migration again.")
-		case err == nil && revErr == nil:
-			log.Info("Migration of ENI datapath successful")
+		case migrated > 0 && err == nil && revErr == nil:
+			log.Infof("Migration of ENI datapath successful, %d rules were migrated", migrated)
 		}
 	}
 
diff --git a/pkg/datapath/linux/routing/migrate.go b/pkg/datapath/linux/routing/migrate.go
index a179ffb0c..3a1936a26 100644
--- a/pkg/datapath/linux/routing/migrate.go
+++ b/pkg/datapath/linux/routing/migrate.go
@@ -24,14 +24,16 @@ import (
 )
 
 // MigrateENIDatapath migrates the egress rules inside the Linux routing policy
-// database (rpdb) for ENI IPAM mode. Because this migration must be
-// all-or-nothing, otherwise breaking connectivity, this function returns two
-// errors: one for the overall migration and one while reverting the successful
-// operation during the migration. For example, if there are 3 egress rules to
-// delete and 2 successfully get deleted, but the last fails, then what will be
-// returned is the error that occurred while deleting the last rule, and any
-// error that occurred while reverting the 2 rules that were successfully
-// deleted (meaning adding those 2 rules).
+// database (rpdb) for ENI IPAM mode. It will return the number of rules that
+// were successfully migrated. Because this migration must be all-or-nothing,
+// otherwise breaking connectivity, this function also returns two errors: one
+// for the overall migration and one while reverting the successful operation
+// during the migration. For example, if there are 3 egress rules to delete and
+// 2 successfully get deleted, but the last fails, then what will be returned
+// is the error that occurred while deleting the last rule, and any error that
+// occurred while reverting the 2 rules that were successfully deleted (meaning
+// adding those 2 rules). In an error case, zero will always be returned as the
+// number of rules actually migrated would be zero.
 //
 // We rely on the ability to fetch the CiliumNode resource because we need to
 // fetch the number associated with the ENI device. The CiliumNode resource
@@ -47,11 +49,10 @@ import (
 // ENI has an ifindex that equals the main routing table number (253-255),
 // causing the rules and routes to be created using the wrong table ID, which
 // could end up blackholing most traffic on the node.
-func MigrateENIDatapath(getter interfaceDB) (error, error) {
+func MigrateENIDatapath(getter interfaceDB) (int, error, error) {
 	rules, err := netlink.RuleList(netlink.FAMILY_V4)
 	if err != nil {
-		return fmt.Errorf("failed to list rules, cannot migrate ENI routing table: %w", err),
-			nil
+		return 0, fmt.Errorf("failed to list rules, cannot migrate ENI routing table: %w", err), nil
 	}
 
 	// This operation will be done over two phases. If either fails, we will
@@ -64,6 +65,9 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 		// migrating.
 		revertStack revert.RevertStack
 
+		// Number of rules successfully migrated. On errors, this will be zero.
+		migrated int
+
 		// Error while migrating and error while reverting, respectively.
 		overallErr, revertErr error
 	)
@@ -130,7 +134,7 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
 		}
 
-		return overallErr, revertErr
+		return 0, overallErr, revertErr
 	}
 
 	for _, r := range rules {
@@ -180,6 +184,10 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 				break
 			}
 
+			// We count the successful cases here because it means we have
+			// passed both phases of the migration.
+			migrated++
+
 		case linux_defaults.RulePriorityEgressv2:
 			// Downgrade case. Unsupported for now.
 		}
@@ -191,10 +199,10 @@ func MigrateENIDatapath(getter interfaceDB) (error, error) {
 			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
 		}
 
-		return overallErr, revertErr
+		return 0, overallErr, revertErr
 	}
 
-	return nil, nil
+	return migrated, nil, nil
 }
 
 // retrieveTableIDFromIfIndex computes the correct table ID based on the
diff --git a/pkg/datapath/linux/routing/migrate_test.go b/pkg/datapath/linux/routing/migrate_test.go
index 5f6cc2f2a..7a36176e1 100644
--- a/pkg/datapath/linux/routing/migrate_test.go
+++ b/pkg/datapath/linux/routing/migrate_test.go
@@ -116,9 +116,10 @@ func (m *MigrateSuite) assertRoutingTable(c *C) {
 	}
 
 	// (2) Make the call to modifying the routing table.
-	err, revErr := MigrateENIDatapath(m)
+	migrated, err, revErr := MigrateENIDatapath(m)
 	c.Assert(err, IsNil)
 	c.Assert(revErr, IsNil)
+	c.Assert(migrated, Equals, 1)
 
 	routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
 		Table: index,
```